### PR TITLE
fix deadlock in mcp server

### PR DESCRIPTION
**Pull request by ayj**
The integration point between the MCP server and snapshot used a
single channel (per-client). This could lead to a deadlock during
configuration pushes.

Address the immediate issues by introduce a dedicated buffered channel
per type.

Longer term we may consider storing pending config pushes in a ring
buffer and only using the channel for indicating when config is ready
to be sent.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -760,43 +760,6 @@ workflows:
       - depupdate
       - build
       - test
-      - e2e-simple:
```
Comment (ayj) [inline .circleci/config.yml]: Temporary change to reduce load on circleci while I debug the flaky MCP e2e tests.

File: pkg/mcp/server/server.go [hunk 2]
```diff
@@ -198,9 +202,19 @@ func (s *Server) StreamAggregatedResources(stream mcp.AggregatedMeshConfigServic
...
 	go con.receive()
 
+	// fan-in per-type response channels into single response channel for the select loop below.
+	responseC := make(chan *WatchResponse)
+	for _, ch := range con.watches {
+		go func(in chan *WatchResponse) {
+			for v := range in {
+				responseC <- v
+			}
+		}(ch.responseC)
```
Comment (hzxuzhonghu) [inline pkg/mcp/server/server.go]: can this solve the issue completely ?

File: galley/pkg/runtime/strategy.go [hunk 3]
```diff
@@ -102,24 +102,28 @@ func (s *publishingStrategy) onTimer() {
...
 			// a deadlock here. Avoid the deadlock by going through the timer loop again.
 			scope.Warnf("Unable to publish to the channel, resetting the timer again to avoid deadlock")
 		}
 	}
 
-	s.timer.Reset(s.timerFrequency)
+	if published {
+		s.timer = nil
+	} else {
+		s.timer.Reset(s.timerFrequency)
```
[the remark this answers is not in the document]
Comment (ayj) [reply galley/pkg/runtime/strategy.go]: `onTimer()` is only invoked when the timer expires via `time.AfterFunc()`.

File: pkg/mcp/server/server.go (hunk 2 above)
Comment (ayj) [reply pkg/mcp/server/server.go]: AFAICT, the root cause of the deadlock was sharing a single channel across watches (per type_url).   The request processing path for one type could become blocked if another type's response push was in flight.

File: pkg/mcp/server/server.go [hunk 4]
```diff
@@ -198,9 +202,19 @@ func (s *Server) StreamAggregatedResources(stream mcp.AggregatedMeshConfigServic
 	defer con.close()
 	go con.receive()
 
+	// fan-in per-type response channels into single response channel for the select loop below.
+	responseChan := make(chan *WatchResponse)
+	for _, ch := range con.watches {
```
Comment (ozevren) [inline pkg/mcp/server/server.go]: nit: ch => watch or w.

**Pull request merged**